drivers: regulator: add support for AXP2101 power management IC

lfelten opened the pull request:
Add initial support for the AXP2101 power management IC from X-powers

gmarull commented on dts/bindings/regulator/x-powers,axp2101-regulator.yaml:
```diff
@@ -0,0 +1,78 @@
+  For example:
+
+  i2c {
+    pmic@68 {
+      reg = <0x68>;
+      ...
+      regulators {
```
you are missing a top-level compatible here, this is a MFD device. It's ok to just add support for regulators now, but DT layout needs to be designed according to how device is.

gmarull commented on tests/drivers/build_all/regulator/i2c.dtsi:
```diff
@@ -76,6 +76,30 @@ axp192@4 {
 };
 
+axp2101@7 {
+	compatible = "x-powers,axp2101";
```
where's this?

gmarull commented on drivers/regulator/regulator_axp2101.c:
```diff
@@ -0,0 +1,463 @@
+
+	ret = i2c_reg_update_byte_dt(&config->i2c, config->desc->enable_reg,
+				     config->desc->enable_mask, 0u);
+	if (ret != 0) {
```
```suggestion
	if (ret < 0) {
```

others as well

lfelten replied:
Hi @gmarull 
thanks, I changed the if statements

gmarull commented on drivers/regulator/regulator_axp2101.c:
```diff
@@ -0,0 +1,463 @@
+	const struct regulator_axp2101_desc *desc;
+	const struct i2c_dt_spec i2c;
+
+	LOG_INSTANCE_PTR_DECLARE(log);
```
Is instance level logging really needed? Very few drivers do this, and, in most cases, I'd say there's only a single regulator instance on a system.

lfelten replied:
Hi @gmarull,
I remove the per instance logging.

kartben commented:
This is a duplicate of #78098?

lfelten commented:
> This is a duplicate of #78098?

Indeed, sorry, didn't check before submitting this one.

lfelten commented:
@kartben @gmarull @soburi 
I compared the drivers, they have similar functionality except that ranranff added the work mode setting.
Should I add the workmode setting here or fix ranranffs driver?
(why can't I mention @ranranff here?)

soburi commented:
@lfelten 

Your and Ranranff's codes are obviously similar because they are copied from regulator_axp192.c. Since the implementation is almost the same, and only the definition is different, I think it is preferable to support it as an extension of regulator_axp192.c.

I tried making something based on your code.

https://github.com/soburi/zephyr/commit/6c74c9dd58c320bae97c58415aee1402255f5b4e
(And one previous commit.)

I have yet to test it, but I hope it helps.

lfelten commented:
@soburi 
Thanks!
I have a patch that adds the sets the workmode for this driver (#82474).
I can test it on one board, but that only uses some DC/DC and LDOs.

> @lfelten
> 
> Your and Ranranff's codes are obviously similar because they are copied from regulator_axp192.c. Since the implementation is almost the same, and only the definition is different, I think it is preferable to support it as an extension of regulator_axp192.c.
> 
> I tried making something based on your code.
> 
> [soburi@6c74c9d](https://github.com/soburi/zephyr/commit/6c74c9dd58c320bae97c58415aee1402255f5b4e) (And one previous commit.)
> 
> I have yet to test it, but I hope it helps.

soburi commented on drivers/regulator/regulator_axp2101.c:
```diff
@@ -0,0 +1,460 @@
+/*
```
Please unify your implementation with axp192.
(Make sure to support axp2101 in the regulator_axp192.c file)
This device is very feature-rich, and duplicating drivers at every layer will make maintenance a pain.

lfelten committed:
drivers: regulator: add support for AXP2101 power management IC

Add initial support for the AXP2101 power management IC from X-powers.
Remark: only DC/DC1 and ALDO have been tested on real hardware.

Co-authored-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>
Co-authored-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

Signed-off-by: Lothar Felten <lothar.felten@gmail.com>

lfelten commented:
@soburi 
The logging was still per instance. @gmarull asked to remove this as usually there is only one PMIC on a board/design.
The tests on my board were successful.

soburi reviewed (dismissed):
LGTM.

It would help if you still addressed pointed by @gmarull https://github.com/zephyrproject-rtos/zephyr/pull/82474#discussion_r1870875175.

lfelten commented:
> LGTM.
> 
> It would help if you still addressed pointed by @gmarull [#82474 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/82474#discussion_r1870875175).

Thanks!
Sorry, that got lost while merging. I added a new version (this is the only change).

soburi commented on dts/bindings/regulator/x-powers,axp2101-regulator.yaml:
```diff
@@ -0,0 +1,135 @@
+
+  &i2c0 {
+      axp2101: axp2101@34 {
+          compatible = "x-powers,axp2101";
```
Bit weird indentation.

soburi commented:
@gmarull 
Could you take a look if you have a while, please?

gmarull commented on drivers/mfd/Kconfig.axp192:
```diff
@@ -4,7 +4,7 @@
 config MFD_AXP192
 	bool "AXP192 PMIC multi-function device driver"
```
please also align Kconfig names/prompts, otherwise it may be misleading

lfelten replied:
Do you prefer the suffix  _AXP192_AXP2101 or _AXP192_2101 for the Kconfig names?

gmarull commented on drivers/mfd/mfd_axp192.c:
```diff
@@ -3,18 +3,49 @@
 
 LOG_MODULE_REGISTER(mfd_axp192, CONFIG_MFD_LOG_LEVEL);
 
+#define AXP_MODEL_COND_11
+#define AXP_MODEL_COND(cond_both_192_2101, cond_192, cond_2101)                                    \
+	COND_CODE_1(UTIL_CAT(UTIL_CAT(AXP_MODEL_COND_,                                             \
+				      CONFIG_DT_HAS_X_POWERS_AXP192_ENABLED),                      \
+			     CONFIG_DT_HAS_X_POWERS_AXP2101_ENABLED),                              \
+		   (cond_both_192_2101),                                                           \
+		   (COND_CODE_1(CONFIG_DT_HAS_X_POWERS_AXP192_ENABLED, (cond_192), (cond_2101))))
+
+#define REG_CHIP_ID(config)                                                                        \
+	AXP_MODEL_COND(config->reg_chip_id, AXP192_REG_CHIP_ID, AXP2101_REG_CHIP_ID)
+#define CHIP_ID(config) AXP_MODEL_COND(config->chip_id, AXP192_CHIP_ID, AXP2101_CHIP_ID)
+#define VBUS_CFG_REG(config)                                                                       \
+	AXP_MODEL_COND(config->vbus_config_reg, AXP192_VBUS_CFG_REG, AXP2101_VBUS_CFG_REG)
+#define VBUS_CFG_VAL_VBUSEN_DISABLE(config)                                                        \
+	AXP_MODEL_COND(config->val_vbusen_disable, AXP192_VBUS_CFG_VAL_VBUSEN_DISABLE,             \
+		       AXP2101_VBUS_CFG_VAL_VBUSEN_DISABLE)
+
+struct mfd_axp192_config {
+	struct i2c_dt_spec i2c;
+#ifdef CONFIG_DT_HAS_X_POWERS_AXP192_ENABLED
+	bool vbusen_disable;
+#endif
+#if CONFIG_DT_HAS_X_POWERS_AXP192_ENABLED && CONFIG_DT_HAS_X_POWERS_AXP2101_ENABLED
+	uint8_t reg_chip_id;
+	uint8_t vbus_config_reg;
+	uint8_t chip_id;
+	uint8_t val_vbusen_disable;
+#endif
+};
```
this is confusing and driver unreadable with these changes. just a simple example:

```
#if CONFIG_DT_HAS_X_POWERS_AXP192_ENABLED && CONFIG_DT_HAS_X_POWERS_AXP2101_ENABLED
```

does that mean I now need to have both regulators for the driver to have this section?
Please, refactor to something that makes sense.

soburi replied:
I did this because I rarely use these two simultaneously, and I don't want to make the table bloated in the typical case (because if there is only one, it can be inlined), but it is a bit roundabout.
If I were to fix it, I would modify it so that this table is always used. Is that OK?

soburi replied:
@lfelten 

Please apply this patch.
https://github.com/soburi/zephyr/commit/2a0f105f1db074b14f8ea539870704c9254ae7a5

lfelten replied:
Patch applied, thanks.
@soburi do you happen to have a complete datasheet? If so, can you please check register 0x3 / CHIP_ID? It's not documented in my copy.

soburi replied:
This seems to be a mistake on my part. It seems to no CHIP_ID.

kartben replied:
https://github.com/hyx0329/axp2101-rs/blob/dev/src%2Fpmu.rs#L561..L567

gmarull commented on dts/bindings/regulator/x-powers,axp2101-regulator.yaml:
```diff
@@ -0,0 +1,139 @@
+        compatible = "x-powers,axp2101_regulator";
+        status = "okay";
+
+        DCDC1: DCDC1 {
```
nodelabels not needed, status okay not needed, compatible wrong, etc.

lfelten replied:
Reverting to old version without details.
The "enum: [0, 1]" is used by silabs,series2-dcdc.yaml. I assume that is ok to use.

gmarull commented on dts/bindings/regulator/x-powers,axp2101-regulator.yaml:
```diff
@@ -0,0 +1,139 @@
+
+        DCDC1: DCDC1 {
+          regulator-name = "DCDC1";
+          regulator-min-microvolt = <1500000>;
+          regulator-max-microvolt = <3400000>;
```
why is this in the example?

lfelten replied:
removed.

lfelten commented:
x-powers,axp2101-regulator.yaml updated.

soburi commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -167,29 +159,283 @@ __maybe_unused static const struct regulator_axp192_desc ldo3_desc = {
+};
+
+static const struct linear_range axp2101_dcdc2_ranges[] = {
+	LINEAR_RANGE_INIT(500000U, 10000U, 0x00U, 0x46U),
+	LINEAR_RANGE_INIT(1220000U, 20000U, 0x00U, 0x10U),
```
LINEAR_RANGE_INIT(500000U, 10000U, 0, 70),
       LINEAR_RANGE_INIT(1220000U, 20000U, 71, 87),

lfelten replied:
corrected, thanks!

soburi commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -167,29 +159,283 @@ __maybe_unused static const struct regulator_axp192_desc ldo3_desc = {
+	.enable_mask = 0x02U,
+	.enable_val = 0x02U,
+	.vsel_reg = AXP2101_REG_DCDC2_VOLTAGE,
+	.vsel_mask = 0x3FU,
```
.vsel_mask = 0x7FU,

lfelten replied:
corrected, thanks!

soburi commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -167,29 +159,283 @@ __maybe_unused static const struct regulator_axp192_desc ldo3_desc = {
+};
+
+static const struct linear_range axp2101_dcdc3_ranges[] = {
+	LINEAR_RANGE_INIT(500000U, 10000U, 0x00U, 0x46U),
+	LINEAR_RANGE_INIT(1220000U, 20000U, 0x00U, 0x10U),
+	LINEAR_RANGE_INIT(1600000U, 100000U, 0x00U, 0x12U),
```
LINEAR_RANGE_INIT(500000U, 10000U, 0, 69),
       LINEAR_RANGE_INIT(1200000U, 20000U, 70, 87),
       LINEAR_RANGE_INIT(1540000U, 60000U, 88, 88),
       LINEAR_RANGE_INIT(1600000U, 100000U, 89, 118),

lfelten replied:
@soburi I've changed it to:
	LINEAR_RANGE_INIT(500000U, 10000U, 0U, 70U),
	LINEAR_RANGE_INIT(1220000U, 20000U, 71U, 87U),
	LINEAR_RANGE_INIT(1600000U, 100000U, 88U, 106U),
Datasheet excerpt:
![image](https://github.com/user-attachments/assets/7af58feb-7c05-4c88-8b3e-c6fb5197a0b7)
Please review

soburi commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -167,29 +159,283 @@ __maybe_unused static const struct regulator_axp192_desc ldo3_desc = {
+	.enable_mask = 0x04U,
+	.enable_val = 0x04U,
+	.vsel_reg = AXP2101_REG_DCDC3_VOLTAGE,
+	.vsel_mask = 0x3FU,
```
.vsel_mask = 0x7FU,

lfelten replied:
corrected, thanks!

soburi commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -167,29 +159,283 @@ __maybe_unused static const struct regulator_axp192_desc ldo3_desc = {
+};
+
+static const struct linear_range axp2101_dcdc4_ranges[] = {
+	LINEAR_RANGE_INIT(500000U, 10000U, 0x0, 0x46),
+	LINEAR_RANGE_INIT(1220000U, 20000U, 0x47, 0x66),
```
LINEAR_RANGE_INIT(500000U, 10000U, 0, 69),
       LINEAR_RANGE_INIT(1200000U, 20000U, 70, 102),

lfelten replied:
@soburi 
changed to 
	LINEAR_RANGE_INIT(500000U, 10000U, 0U, 70U),
	LINEAR_RANGE_INIT(1220000U, 20000U, 71U, 102U),

![image](https://github.com/user-attachments/assets/cf2d77c8-a425-4e50-88d8-cd8e4545c618)

Please review

lfelten commented:
Rebased to main, the SMP related build failure was fixed (thanks @sylvioalves !)
Now there is a crypto related issue.

soburi commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -350,44 +601,63 @@ static int regulator_axp192_init(const struct device *dev)
+	REGULATOR_AXP2101_DEFINE_COND(node, aldo4)                                                 \
+	REGULATOR_AXP2101_DEFINE_COND(node, bldo1)                                                 \
+	REGULATOR_AXP2101_DEFINE_COND(node, bldo2)                                                 \
+	REGULATOR_AXP2101_DEFINE_COND(node, cpuldo)                                                \
```
REGULATOR_AXP2101_DEFINE_COND(node, cpusldo)

lfelten replied:
Applied, thanks!

lfelten commented:
I will rebase when #83822 is merged

lfelten commented:
Rebased to main now that #82813 was merged

soburi commented:
@lfelten 

From what I've tried, there seem to be some general issues with using ranges, and the values ​​set in the registers seem strange. Have you checked this?

lfelten commented:
> @lfelten
> 
> From what I've tried, there seem to be some general issues with using ranges, and the values ​​set in the registers seem strange. Have you checked this?

I have checked some of the LDOs on my board using a multimeter. So the settings do set the correct voltage on the LDOs I tested (need to check which ones, it has been a few weeks). I don't know why that should be ranges, though.

lfelten commented:
@soburi I still have no hardware that uses a DC/DC besides DCDC1, sorry. 
Here is a summary of the changes:

- converted all ranges to unsigned decimal notation
- corrected ranges for dcdc2, dcdc3, dcdc4
- corrected vsel masks for dcdc2, dcdc3, dcdc4, dcdc5
- consolidated ranges for ald1/2/3/4 and bldo1/2 into abldox (range 
- rebased to main

soburi commented:
@lfelten 

> I still have no hardware that uses a DC/DC besides DCDC1, sorry.

Please note this in the commit message about this.

soburi reviewed (dismissed):
LGTM

soburi commented:
@gmarull 
Could you take a look again?

gmarull commented on include/zephyr/drivers/mfd/axp2101.h:
```diff
@@ -0,0 +1,39 @@
+#ifndef ZEPHYR_INCLUDE_DRIVERS_MFD_AXP2101_H_
+#define ZEPHYR_INCLUDE_DRIVERS_MFD_AXP2101_H_
+
+#define AXP2101_REG_CHIP_ID           0x03U
+#define AXP2101_REG_CHGLED            0x69U
+#define AXP2101_REG_DCDC12345_CONTROL 0x80U
+#define AXP2101_REG_DCDCS_PWM_CONTROL 0x81U
+#define AXP2101_REG_DCDC1_VOLTAGE     0x82U
+#define AXP2101_REG_DCDC2_VOLTAGE     0x83U
+#define AXP2101_REG_DCDC3_VOLTAGE     0x84U
+#define AXP2101_REG_DCDC4_VOLTAGE     0x85U
+#define AXP2101_REG_DCDC5_VOLTAGE     0x86U
+#define AXP2101_REG_LDOGRP1_CONTROL   0x90U
+#define AXP2101_REG_LDOGRP2_CONTROL   0x91U
+#define AXP2101_REG_ALDO1_VOLTAGE     0x92U
+#define AXP2101_REG_ALDO2_VOLTAGE     0x93U
+#define AXP2101_REG_ALDO3_VOLTAGE     0x94U
+#define AXP2101_REG_ALDO4_VOLTAGE     0x95U
+#define AXP2101_REG_BLDO1_VOLTAGE     0x96U
+#define AXP2101_REG_BLDO2_VOLTAGE     0x97U
+#define AXP2101_REG_CPUSLDO_VOLTAGE   0x98U
+#define AXP2101_REG_DLDO1_VOLTAGE     0x99U
+#define AXP2101_REG_DLDO2_VOLTAGE     0x9AU
+
```
why is this all exposed in a public header?

lfelten replied:
Hello @gmarull, thanks for your review.
This is done in the style of the original axp-192 driver.
From axp192.h:
`
#ifdef __cplusplus
extern "C" {
#endif

/* Regulator output control registers */
#define AXP192_REG_EXTEN_DCDC2_CONTROL   0x10U
#define AXP192_REG_DCDC123_LDO23_CONTROL 0x12U
#define AXP192_REG_DCDC2_VOLTAGE         0x23U
#define AXP192_REG_DCDC2_SLOPE           0x25U
#define AXP192_REG_DCDC1_VOLTAGE         0x26U
#define AXP192_REG_DCDC3_VOLTAGE         0x27U
#define AXP192_REG_LDO23_VOLTAGE         0x28U
#define AXP192_REG_DCDC123_WORKMODE      0x80U
#define AXP192_REG_GPIO0_CONTROL         0x90U
#define AXP192_REG_LDOIO0_VOLTAGE        0x91U
...
`

We can probably move all those defines to the top of .c file, if you prefer this.

gmarull replied:
public header should expose stuff that it's needed by clients, nothing else.

lfelten replied:
I moved the defines for AXP2101 and AXP192.

gmarull commented on drivers/regulator/regulator_axp192.c:
```diff
@@ -18,9 +17,7 @@
 #include <zephyr/logging/log_instance.h>
 #include <zephyr/drivers/mfd/axp192.h>
 
-LOG_MODULE_REGISTER(regulator_axp192, CONFIG_REGULATOR_LOG_LEVEL);
```
why moving this?

lfelten replied:
Hi, moved it back, sorry.

lfelten commented on drivers/mfd/mfd_axp192.c:
```diff
@@ -605,16 +615,29 @@ int mfd_axp192_gpio_write_port(const struct device *dev, uint8_t value, uint8_t
 
-DT_INST_FOREACH_STATUS_OKAY(MFD_AXP192_DEFINE);
+DT_FOREACH_STATUS_OKAY_VARGS(x_powers_axp192, MFD_AXP192_DEFINE, 192);
+DT_FOREACH_STATUS_OKAY_VARGS(x_powers_axp2101, MFD_AXP192_DEFINE, 2101);
```
I hope I understood it right: I renamed MFD_AXP192_DEFINE to MFD_AXP192__AXP2101_DEFINE to indicate it applies to axp192 and axp2101.

lfelten commented:
Rebased to current main branch
@gmarull: your requested changes were added.
I also fixed the register defines (#ifdef... clause) for AXP2101
